lib/resolvers/parse/ast/literal.js

```diff
@@ -0,0 +1,61 @@
+const { Kind } = require('graphql')
+
+const _getTypeFrom_fieldOr_arg = _field => {
+  let { type } = _field
+  while (type.ofType) type = type.ofType
+  return type
+}
+
+const _getTypeFrom_fields = (_fields, path, index = 0) => {
+  const { name } = path[index++]
+
+  const _field = _fields[name]
+  const type = _getTypeFrom_fieldOr_arg(_field)
+
+  // If type has the parseLiteral function it is a scalar type -> leaf -> end of path
+  if (type.parseLiteral) return type
+
+  const next = path[index]
+  // Is the next path element an argument? If yes, follow the argument
+  if (next.kind === Kind.ARGUMENT) {
+    const arg = _field.args.find(a => a.name === next.name)
+    const type = _getTypeFrom_fieldOr_arg(arg)
+
+    // If type has the parseLiteral function it is a scalar type -> leaf -> end of path
+    // This case occurs when the argument itself is a scalar type, e.g. Books(top: 1)
+    if (type.parseLiteral) return type
+
+    return _getTypeFrom_fields(type.getFields(), path, index + 1)
+  }
+
+  return _getTypeFrom_fields(type.getFields(), path, index)
+}
+
+const _pathKinds = [Kind.FIELD, Kind.ARGUMENT, Kind.OBJECT_FIELD]
+const _isPathKind = _pathKinds.includes.bind(_pathKinds)
+const _simplifiedPath = (node, ancestors) =>
+  ancestors
+    .concat(node)
+    .filter(e => _isPathKind(e.kind))
+    .map(e => ({ kind: e.kind, name: e.name.value }))
```

Comment:
When defining a function inside a function definition. It is actually created on each call. Which means that `V8` will avoid optimizing it as it is required to call a function multiple times before it will be optimized. The number of times a function has to be called before it being optimized depends on the function body size. So for this instance it is really not important.

Especially for this scenario the real optimization would be to use a `for` loop instead of the `Array` functions. With the possibility to not copy the array three times. As `concat`, `filter` and `map` all copy the `Array`.

```suggestion
const _pathKinds = [Kind.FIELD, Kind.ARGUMENT, Kind.OBJECT_FIELD]
const _isPathKind = e => _pathKinds.includes(e.kind)
const _simplifiedPath = (node, ancestors) =>
  ancestors
    .concat(node)
    .filter(_isPathKind)
    .map(e => ({ kind: e.kind, name: e.name.value }))
```

Reply:
I see, thanks for the performance recommendations!

So, would you instead recommend an implementation such as this?

```javascript
const _simplifiedPath = (node, ancestors) => {
  const path = []

  for (const ancestor of ancestors) {
    if (!_isNodePathKind(ancestor)) continue
    path.push({ kind: ancestor.kind, name: ancestor.name.value })
  }

  if (_isNodePathKind(node)) path.push({ kind: node.kind, name: node.name.value })

  return path
}
```

Or would you say the performance gain is negligible and the readability/brevity of the array functions is preferred?

Or are there perhaps even further performance recommendations in the snippet above, such as iterating using an index vs `of` or avoiding `array.push`?